### PR TITLE
Telemetry: Changes to metric spec and enhancements to lvm exporter 

### PR DESCRIPTION
**Pull request by NajmudheenCT**
**What this PR does / why we need it**:
1. Changes to the metricspec structure. Removed unsused fields , renamed associatorMap to Labels, added new field Aggr_type
2. Update Lables map in driver.
3. In exporter, use Labels to update metrics lables for prometheus  
4.Adding Disk metrics support in lvm exporter
5.Reading resource list fro which metrics to be collected from a config file (resources.yaml)

**Special notes for your reviewer**:
NA

### CODE REVIEW

**Comment by NajmudheenCT** [comment]
@xing-yang @leonwanghui @rhsakarpos @skdwriting

File: pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go [hunk 1]
```diff
@@ -28,45 +30,77 @@ import (
...
-			volumeLabel, nil,
+			labelKeys, nil,
 		),
-		Utilization: prometheus.NewDesc("OpensSDS_Volume_Utilization_prcnt",
+		VolumeUtilization: prometheus.NewDesc("OpensSDS_Volume_Utilization_prcnt",
 			"Shows Utilization in percentage",
-			volumeLabel, nil,
+			labelKeys, nil,
+		),
+		DiskIOPS: prometheus.NewDesc("OpensSDS_Volume_IOPS_tps",
```
Comment (leonwanghui) [inline pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go]: What do you mean here `OpensSDS`?

File: pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go [hunk 2]
```diff
@@ -28,45 +30,77 @@ import (
...
+			"Shows IOPS",
+			labelKeys, nil,
+		),
+		DiskReadThroughput: prometheus.NewDesc("OpensSDS_Disk_ReadThroughput_KBs",
```
Comment (leonwanghui) [inline pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go]: Same here

File: pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go [hunk 3]
```diff
@@ -77,12 +111,27 @@ func newLvmCollector() *lvmCollector {
...
+	ch <- c.DiskUtilization
+}
+
+type Config struct {
+	Type    string   `type`
+	Devices []string `devices`
+}
+
+type Configs struct {
+	Cfgs []Config `resources`
```
Comment (leonwanghui) [inline pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go]: s/[]Config/[]*Config

File: pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go [hunk 4]
```diff
@@ -94,33 +143,74 @@ func (c *lvmCollector) Collect(ch chan<- prometheus.Metric) {
 	//Implement logic here to determine proper metric value to return to prometheus
 	//for each descriptor
 	metricList := []string{"IOPS", "ReadThroughput", "WriteThroughput", "ResponseTime", "ServiceTime", "UtilizationPercentage"}
-	//Todo : Need to read list from a config file
-	volumeList := []string{"sda", "loop0"}
+	filename := "resources.yaml"
```
Comment (leonwanghui) [inline pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go]: Please use a global constant here

File: pkg/controller/metrics/exporters/lvm_exporter/resources.yaml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+resources:
+  - type: volume
+    devices:
+      - volume-e138cb0d-375d-4118-aee6-2ce6c12f17a6
+
```
Comment (leonwanghui) [inline pkg/controller/metrics/exporters/lvm_exporter/resources.yaml]: Please remove the blank line

File: pkg/model/metric.go [hunk 6]
```diff
@@ -71,13 +71,13 @@ type MetricSpec struct {
 	  the associated component type as key and component name as value
 	  Example: associator[pool]=pool1 */
 
-	Associator map[string]string `json:"associator,omitempty"`
+	Labels map[string]string `json:"associator,omitempty"`
```
Comment (leonwanghui) [inline pkg/model/metric.go]: Please change the json label here.

File: pkg/model/metric.go [hunk 7]
```diff
@@ -71,13 +71,13 @@ type MetricSpec struct {
...
 
 	// Following fields can be used to form a unique metric name
 
 	// source -\> Node/Dock
 
-	Source string `json:"source,omitempty"`
+	//Source string `json:"source,omitempty"`
```
Comment (leonwanghui) [inline pkg/model/metric.go]: Please remove this line

File: pkg/model/metric.go [hunk 8]
```diff
@@ -91,11 +91,8 @@ type MetricSpec struct {
 
 	Unit string `json:"unit,omitempty"`
 
-	// is aggregated
-
-	IsAggregated bool `json:"isAggregated,omitempty"`
-
-	// aggr_type-\> Can be used to determine Total/Sum/Avg etc
+	// Can be used to determine Total/Avg etc
+	Aggr_type string `json:"unit,omitempty"`
```
Comment (leonwanghui) [inline pkg/model/metric.go]: Please keep the json name same with field name

File: pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go (hunk 1 above)
Comment (NajmudheenCT) [reply pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go]: @leonwanghui , Earlier we thought of naming with OpenSDS to identify metrics collected using opensds modules, but prometheus has naming stranded of using exporter identity to start the metric name, hence using 'lvm' instead if 'OpenSDS'.

**Comment by NajmudheenCT** [comment]
> Please use `goimports` to format the code style.

Thnaks, I ran 'goimports' and updated.

File: pkg/model/metric.go [hunk 9]
```diff
@@ -91,11 +87,8 @@ type MetricSpec struct {
 
 	Unit string `json:"unit,omitempty"`
 
-	// is aggregated
-
-	IsAggregated bool `json:"isAggregated,omitempty"`
-
-	// aggr_type-\> Can be used to determine Total/Sum/Avg etc
+	// Can be used to determine Total/Avg etc
+	Aggr_type string `json:"aggr_type,omitempty"`
```
Comment (leonwanghui) [inline pkg/model/metric.go]: Please use `AggrType` as the field name.

File: pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go [hunk 10]
```diff
@@ -133,6 +222,7 @@ func validateCliArg(arg1 string) string {
 	}
 	return arg1
 }
+
 // main function for lvm exporter
 // lvm exporter is a independent process which user can start if required
 func main() {
```
Comment (leonwanghui) [inline pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go]: I don't know why there is a main method here.

File: pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go [hunk 11]
```diff
@@ -15,58 +15,95 @@ package main
...
+	//Disk metrics
+	DiskIOPS            *prometheus.Desc
+	DiskReadThroughput  *prometheus.Desc
+	DiskWriteThroughput *prometheus.Desc
+	DiskResponseTime    *prometheus.Desc
+	DiskServiceTime     *prometheus.Desc
+	DiskUtilization     *prometheus.Desc
 }
 
 /* rr */
```
Comment (xing-yang) [inline pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go]: What does this mean?  Please remove it.

File: pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go [hunk 12]
```diff
@@ -15,58 +15,95 @@ package main
...
 //constructor for lvm collector that
 //initializes every descriptor and returns a pointer to the collector
```
Comment (xing-yang) [inline pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go]: Add a space after "//"

File: pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go [hunk 13]
```diff
@@ -94,33 +146,70 @@ func (c *lvmCollector) Collect(ch chan<- prometheus.Metric) {
...
+	if err != nil {
+		log.Fatal("file "+DefaultConfigFile+"can't read", err)
+	}
+	var config Configs
+	err1 := yaml.Unmarshal(source, &config)
+	if err1 != nil {
+		log.Fatalf("error: %v", err)
+	}
+
+	//volumeList := []string{"sda", "loop0"}
```
Comment (xing-yang) [inline pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go]: Remove commented out code.

File: pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go [hunk 14]
```diff
@@ -94,33 +146,70 @@ func (c *lvmCollector) Collect(ch chan<- prometheus.Metric) {
...
+					}
+				}
+			}
 
+		case "disk":
+			for _, volume := range resource.Devices {
+				metricArray, _ := metricDriver.CollectMetrics(metricList, volume)
+				for _, metric := range metricArray {
+					lableVals := []string{metric.Labels["device"]}
+					//unitLabel := "Unit:"+metric.Unit
```
Comment (xing-yang) [inline pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go]: Remove

File: pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go (hunk 10 above)
Comment (NajmudheenCT) [reply pkg/controller/metrics/exporters/lvm_exporter/lvm_exporter.go]: LVM Exporter is an independent process which user starts on demand. User can either use lvm exporter or periodic post API . This is why we have main() here.

**Comment by xing-yang** [comment]
Looks good, but need a rebase before it can be merged.

**Comment by NajmudheenCT** [comment]
> Looks good, but need a rebase before it can be merged.

Done

**Comment by xing-yang** [comment]
It still says "This branch is out-of-date with the base branch".

**Pull request merged**